src/TraceEvent/EventPipe/EventPipeEventSourceV1.cs

```diff
@@ -0,0 +1,373 @@
+        {
+            var hdr = InitEventRecord(eventMetadata, threadId, timeStamp, activityId, relatedActivityId, payload);
+            TraceEvent traceEvent = Lookup(hdr);
+            traceEvent.eventRecord = hdr;
+            return traceEvent;
+        }
+
+        private TraceEvent SynthesizeThreadStackWalkEvent(int threadId, byte[] stack)
+        {
+            _header->EventHeader.Size = (ushort)sizeof(TraceEventNativeMethods.EVENT_TRACE_HEADER);
+            _header->EventHeader.Flags = 0;
+            if (pointerSize == 8)
+                _header->EventHeader.Flags |= TraceEventNativeMethods.EVENT_HEADER_FLAG_64_BIT_HEADER;
+            else
+                _header->EventHeader.Flags |= TraceEventNativeMethods.EVENT_HEADER_FLAG_32_BIT_HEADER;
+
+            _header->EventHeader.ThreadId = threadId;
+            _header->EventHeader.ProviderId = SampleProfilerTraceEventParser.ProviderGuid;
+            _header->EventHeader.Id = 1; // StackWalk
+
+            _header->ExtendedDataCount = 0;
+
+            _header->UserDataLength = (ushort)stack.Length;
+            _header->UserData = GCHandle.Alloc(stack, GCHandleType.Pinned).AddrOfPinnedObject();
+
+            TraceEvent traceEvent = Lookup(_header);
+            traceEvent.eventRecord = _header;
+            return traceEvent;
+        }
+
+        private TraceEventNativeMethods.EVENT_RECORD* InitEventRecord(EventMetadata eventMetadata, int threadId, long timeStamp, Guid activityId, Guid relatedActivityId, byte[] payload)
+        {
+            _header->EventHeader.Size = (ushort)sizeof(TraceEventNativeMethods.EVENT_TRACE_HEADER);
+            _header->EventHeader.Flags = 0;
+            if (pointerSize == 8)
+                _header->EventHeader.Flags |= TraceEventNativeMethods.EVENT_HEADER_FLAG_64_BIT_HEADER;
+            else
+                _header->EventHeader.Flags |= TraceEventNativeMethods.EVENT_HEADER_FLAG_32_BIT_HEADER;
+
+            _header->EventHeader.TimeStamp = timeStamp;
+            _header->EventHeader.ProviderId = eventMetadata.ProviderId;
+            _header->EventHeader.Version = (byte)eventMetadata.Version;
+            _header->EventHeader.Level = (byte)eventMetadata.Level;
+            _header->EventHeader.Opcode = GetOpcodeFromEventName(eventMetadata.EventName);
+            _header->EventHeader.Id = (ushort)eventMetadata.EventId;
+            _header->EventHeader.Keyword = eventMetadata.Keywords;
+
+            _header->EventHeader.ActivityId = activityId;
+            _header->ExtendedDataCount = 1;
+            _header->ExtendedData = (TraceEventNativeMethods.EVENT_HEADER_EXTENDED_DATA_ITEM*)Marshal.AllocHGlobal(sizeof(TraceEventNativeMethods.EVENT_HEADER_EXTENDED_DATA_ITEM));
+            _header->ExtendedData->ExtType = TraceEventNativeMethods.EVENT_HEADER_EXT_TYPE_RELATED_ACTIVITYID;
```

Comment:
You don't need to do this extended type thing.  TraceLog already had this problem of how to serialize extended data, and that is what the 'TraceEvent.myBuffer' field gets used for.   See TraceEvent.Clone() for more details, but basically things a TraceEvent understands how to be 'own' its own native memory (It will clean it up too),  You just need to allocate the extra space (for both its payload and if needed RelatedActivityID) and set it up load 'Clone()' does and TraceEvent will decode it properly.